File: app/src/main/java/com/termirust/mobile/controller/ControllerWriterControl.kt
package com.termirust.mobile.controller

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.UUID

sealed interface WriterLeaseState {
    data object None : WriterLeaseState
    data class Requesting(val commandId: UUID) : WriterLeaseState
    data object Held : WriterLeaseState
    data object Busy : WriterLeaseState
    data object Lost : WriterLeaseState
}

enum class PendingInputKind { KEYBOARD, PASTE }

data class PendingInput(
    val commandId: UUID,
    val bytes: ByteArray,
    val kind: PendingInputKind,
) {
    override fun equals(other: Any?): Boolean = other is PendingInput &&
        commandId == other.commandId && kind == other.kind && bytes.contentEquals(other.bytes)
    override fun hashCode(): Int = commandId.hashCode()
}

object ControllerWriterWireCodec {
    const val MAX_INPUT_CHUNK_BYTES = 16 * 1_024
    private val json = Json { encodeDefaults = true; explicitNulls = true }

    fun encodeAcquire(commandId: UUID, sessionGeneration: Long, deadlineMillis: Long, identity: ReadOnlyAttachIdentity) =
        encodeLease("acquire_writer", commandId, sessionGeneration, deadlineMillis, identity)

    fun encodeRelease(commandId: UUID, sessionGeneration: Long, deadlineMillis: Long, identity: ReadOnlyAttachIdentity) =
        encodeLease("release_writer", commandId, sessionGeneration, deadlineMillis, identity)

    fun encodeInput(
        commandId: UUID,
        sessionGeneration: Long,
        deadlineMillis: Long,
        identity: ReadOnlyAttachIdentity,
        bytes: ByteArray,
    ): ByteArray {
        validateEnvelope(sessionGeneration, deadlineMillis, identity)
        require(bytes.size in 1..MAX_INPUT_CHUNK_BYTES)
        return encode(
            commandId,
            sessionGeneration,
            deadlineMillis,
            InputCommand(
                sessionId = identity.sessionId.toString(),
                occupantGeneration = identity.occupantGeneration,
                bytes = bytes.map { it.toInt() and 0xff },
            ),
        )
    }

    fun encodeResize(
        commandId: UUID,
        sessionGeneration: Long,
        deadlineMillis: Long,
        identity: ReadOnlyAttachIdentity,
        viewport: TerminalViewport,
        limits: TerminalLimits = TerminalLimits(),
    ): ByteArray {
        validateEnvelope(sessionGeneration, deadlineMillis, identity)
        limits.validate(viewport)
        return encode(
            commandId,
            sessionGeneration,
            deadlineMillis,
            ResizeCommand(
                sessionId = identity.sessionId.toString(),
                occupantGeneration = identity.occupantGeneration,
                columns = viewport.columns,
                rows = viewport.rows,
            ),
        )
    }

    private fun encodeLease(
        kind: String,
        commandId: UUID,
        sessionGeneration: Long,
        deadlineMillis: Long,
        identity: ReadOnlyAttachIdentity,
    ): ByteArray {
        validateEnvelope(sessionGeneration, deadlineMillis, identity)
        return encode(
            commandId,
            sessionGeneration,
            deadlineMillis,
            LeaseCommand(kind, identity.sessionId.toString(), identity.occupantGeneration),
        )
    }

    private inline fun <reified T> encode(
        commandId: UUID,
        sessionGeneration: Long,
        deadlineMillis: Long,
        command: T,
    ): ByteArray where T : WriterCommand {
        val payload = json.encodeToString(
            WriterEnvelope(
                commandId = commandId.toString(),
                sessionGeneration = sessionGeneration,
                deadlineMillis = deadlineMillis,
                command = command,
            ),
        ).encodeToByteArray()
        require(payload.size <= MAX_CONTROL_PAYLOAD_BYTES)
        return payload
    }

    private fun validateEnvelope(sessionGeneration: Long, deadlineMillis: Long, identity: ReadOnlyAttachIdentity) {
        identity.validate()
        require(sessionGeneration > 0 && deadlineMillis > 0)
    }

    private const val MAX_CONTROL_PAYLOAD_BYTES = 64 * 1_024
}

class WriterControlReducer(val identity: ReadOnlyAttachIdentity) {
    var lease: WriterLeaseState = WriterLeaseState.None
        private set
    var isForeground: Boolean = true
        private set
    var queuedBytes: Int = 0
        private set
    private val queue = ArrayDeque<PendingInput>()

    init { identity.validate() }

    fun beginAcquire(commandId: UUID) {
        require(isForeground && (lease == WriterLeaseState.None || lease == WriterLeaseState.Busy || lease == WriterLeaseState.Lost))
        lease = WriterLeaseState.Requesting(commandId)
    }

    fun finishAcquire(commandId: UUID, applied: Boolean) {
        require(lease == WriterLeaseState.Requesting(commandId))
        lease = if (applied) WriterLeaseState.Held else WriterLeaseState.Busy
    }

    fun markLeaseLost() {
        lease = WriterLeaseState.Lost
        clearPendingInput()
    }

    fun releaseLocally() {
        lease = WriterLeaseState.None
        clearPendingInput()
    }

    fun pasteRequiresConfirmation(bytes: ByteArray): Boolean =
        bytes.size > PASTE_CONFIRMATION_BYTES || bytes.any { it == '\n'.code.toByte() || it == '\r'.code.toByte() }

    fun enqueue(
        bytes: ByteArray,
        kind: PendingInputKind,
        confirmed: Boolean = false,
        commandId: UUID = UUID.randomUUID(),
    ) {
        require(isForeground && lease == WriterLeaseState.Held)
        require(bytes.size in 1..ControllerWriterWireCodec.MAX_INPUT_CHUNK_BYTES)
        require(kind != PendingInputKind.PASTE || confirmed || !pasteRequiresConfirmation(bytes))
        require(queue.size < MAX_QUEUED_CHUNKS)
        require(bytes.size <= MAX_QUEUED_BYTES - queuedBytes)
        queue.addLast(PendingInput(commandId, bytes.copyOf(), kind))
        queuedBytes += bytes.size
    }

    fun removeFirstOrNull(): PendingInput? = queue.removeFirstOrNull()?.also { queuedBytes -= it.bytes.size }

    fun setForeground(foreground: Boolean) {
        isForeground = foreground
        if (!foreground) markLeaseLost()
    }

    private fun clearPendingInput() {
        queue.clear()
        queuedBytes = 0
    }

    companion object {
        const val MAX_QUEUED_CHUNKS = 64
        const val MAX_QUEUED_BYTES = 256 * 1_024
        const val PASTE_CONFIRMATION_BYTES = 4 * 1_024
    }
}

private sealed interface WriterCommand

@Serializable private data class WriterEnvelope<T : WriterCommand>(
    val version: Int = 1,
    @SerialName("command_id") val commandId: String,
    @SerialName("session_generation") val sessionGeneration: Long,
    @SerialName("deadline_millis") val deadlineMillis: Long,
    val command: T,
)

@Serializable private data class LeaseCommand(
    val kind: String,
    @SerialName("session_id") val sessionId: String,
    @SerialName("occupant_generation") val occupantGeneration: Long,
) : WriterCommand

@Serializable private data class InputCommand(
    val kind: String = "input",
    @SerialName("session_id") val sessionId: String,
    @SerialName("occupant_generation") val occupantGeneration: Long,
    val bytes: List<Int>,
) : WriterCommand

@Serializable private data class ResizeCommand(
    val kind: String = "resize",
    @SerialName("session_id") val sessionId: String,
    @SerialName("occupant_generation") val occupantGeneration: Long,
    val columns: Int,
    val rows: Int,
) : WriterCommand

File: TermiRustMobile/Controller/ControllerWriterControl.swift
import Foundation

enum WriterLeaseState: Equatable, Sendable {
    case none
    case requesting(commandID: UUID)
    case held
    case busy
    case lost
}

enum PendingInputKind: Equatable, Sendable {
    case keyboard
    case paste
}

struct PendingInput: Equatable, Sendable {
    let commandID: UUID
    let bytes: Data
    let kind: PendingInputKind
}

enum WriterControlFailure: Error, Equatable, Sendable {
    case invalidTransition
    case capabilityDenied
    case inputEmpty
    case inputChunkTooLarge
    case inputQueueFull
    case inputQueueBytesExceeded
    case pasteConfirmationRequired
    case staleCommand
    case resizeOutOfRange
}

private struct WriterCommandEnvelope<Command: Encodable>: Encodable {
    let version = 1
    let commandId: UUID
    let sessionGeneration: UInt64
    let deadlineMillis: UInt64
    let command: Command

    private enum CodingKeys: String, CodingKey {
        case version
        case commandId = "command_id"
        case sessionGeneration = "session_generation"
        case deadlineMillis = "deadline_millis"
        case command
    }
}

private struct WriterLeaseCommand: Encodable {
    let kind: String
    let sessionId: UUID
    let occupantGeneration: UInt64

    private enum CodingKeys: String, CodingKey {
        case kind
        case sessionId = "session_id"
        case occupantGeneration = "occupant_generation"
    }
}

private struct WriterInputCommand: Encodable {
    let kind = "input"
    let sessionId: UUID
    let occupantGeneration: UInt64
    let bytes: [UInt8]

    private enum CodingKeys: String, CodingKey {
        case kind
        case sessionId = "session_id"
        case occupantGeneration = "occupant_generation"
        case bytes
    }
}

private struct WriterResizeCommand: Encodable {
    let kind = "resize"
    let sessionId: UUID
    let occupantGeneration: UInt64
    let columns: UInt32
    let rows: UInt32

    private enum CodingKeys: String, CodingKey {
        case kind
        case sessionId = "session_id"
        case occupantGeneration = "occupant_generation"
        case columns
        case rows
    }
}

enum ControllerWriterWireCodec {
    static let maxInputChunkBytes = 16 * 1_024

    static func encodeAcquireWriter(
        commandID: UUID,
        sessionGeneration: UInt64,
        deadlineMillis: UInt64,
        identity: ReadOnlyAttachIdentity
    ) throws -> Data {
        try encodeLease(
            kind: "acquire_writer",
            commandID: commandID,
            sessionGeneration: sessionGeneration,
            deadlineMillis: deadlineMillis,
            identity: identity
        )
    }

    static func encodeReleaseWriter(
        commandID: UUID,
        sessionGeneration: UInt64,
        deadlineMillis: UInt64,
        identity: ReadOnlyAttachIdentity
    ) throws -> Data {
        try encodeLease(
            kind: "release_writer",
            commandID: commandID,
            sessionGeneration: sessionGeneration,
            deadlineMillis: deadlineMillis,
            identity: identity
        )
    }

    static func encodeInput(
        commandID: UUID,
        sessionGeneration: UInt64,
        deadlineMillis: UInt64,
        identity: ReadOnlyAttachIdentity,
        bytes: Data
    ) throws -> Data {
        try validateEnvelope(
            sessionGeneration: sessionGeneration,
            deadlineMillis: deadlineMillis,
            identity: identity
        )
        guard !bytes.isEmpty else { throw WriterControlFailure.inputEmpty }
        guard bytes.count <= maxInputChunkBytes else {
            throw WriterControlFailure.inputChunkTooLarge
        }
        return try JSONEncoder().encode(WriterCommandEnvelope(
            commandId: commandID,
            sessionGeneration: sessionGeneration,
            deadlineMillis: deadlineMillis,
            command: WriterInputCommand(
                sessionId: identity.sessionID,
                occupantGeneration: identity.occupantGeneration,
                bytes: Array(bytes)
            )
        ))
    }

    static func encodeResize(
        commandID: UUID,
        sessionGeneration: UInt64,
        deadlineMillis: UInt64,
        identity: ReadOnlyAttachIdentity,
        viewport: TerminalViewportState,
        limits: TerminalLimits = .controllerDefault
    ) throws -> Data {
        try validateEnvelope(
            sessionGeneration: sessionGeneration,
            deadlineMillis: deadlineMillis,
            identity: identity
        )
        do {
            try limits.validate(viewport: viewport)
        } catch {
            throw WriterControlFailure.resizeOutOfRange
        }
        return try JSONEncoder().encode(WriterCommandEnvelope(
            commandId: commandID,
            sessionGeneration: sessionGeneration,
            deadlineMillis: deadlineMillis,
            command: WriterResizeCommand(
                sessionId: identity.sessionID,
                occupantGeneration: identity.occupantGeneration,
                columns: UInt32(viewport.columns),
                rows: UInt32(viewport.rows)
            )
        ))
    }

    private static func encodeLease(
        kind: String,
        commandID: UUID,
        sessionGeneration: UInt64,
        deadlineMillis: UInt64,
        identity: ReadOnlyAttachIdentity
    ) throws -> Data {
        try validateEnvelope(
            sessionGeneration: sessionGeneration,
            deadlineMillis: deadlineMillis,
            identity: identity
        )
        return try JSONEncoder().encode(WriterCommandEnvelope(
            commandId: commandID,
            sessionGeneration: sessionGeneration,
            deadlineMillis: deadlineMillis,
            command: WriterLeaseCommand(
                kind: kind,
                sessionId: identity.sessionID,
                occupantGeneration: identity.occupantGeneration
            )
        ))
    }

    private static func validateEnvelope(
        sessionGeneration: UInt64,
        deadlineMillis: UInt64,
        identity: ReadOnlyAttachIdentity
    ) throws {
        try identity.validate()
        // Early desktop authorities used zero as their exact initial generation.
        guard deadlineMillis > 0 else {
            throw WriterControlFailure.invalidTransition
        }
    }
}

struct WriterControlReducer: Sendable {
    static let maxQueuedChunks = 64
    static let maxQueuedBytes = 256 * 1_024
    static let pasteConfirmationBytes = 4 * 1_024

    let identity: ReadOnlyAttachIdentity
    private(set) var lease: WriterLeaseState = .none
    private(set) var isForeground = true
    private var queue: [PendingInput] = []
    private(set) var queuedBytes = 0

    init(identity: ReadOnlyAttachIdentity) throws {
        try identity.validate()
        self.identity = identity
    }

    mutating func beginAcquire(commandID: UUID) throws {
        guard isForeground,
              lease == .none || lease == .busy || lease == .lost else {
            throw WriterControlFailure.invalidTransition
        }
        lease = .requesting(commandID: commandID)
    }

    mutating func finishAcquire(commandID: UUID, applied: Bool) throws {
        guard lease == .requesting(commandID: commandID) else {
            throw WriterControlFailure.staleCommand
        }
        lease = applied ? .held : .busy
    }

    mutating func markLeaseLost() {
        lease = .lost
        clearPendingInput()
    }

    mutating func releaseLocally() {
        lease = .none
        clearPendingInput()
    }

    func pasteRequiresConfirmation(_ bytes: Data) -> Bool {
        bytes.count > Self.pasteConfirmationBytes
            || bytes.contains(0x0A)
            || bytes.contains(0x0D)
    }

    mutating func enqueue(
        _ bytes: Data,
        kind: PendingInputKind,
        confirmed: Bool = false,
        commandID: UUID = UUID()
    ) throws {
        guard isForeground, lease == .held else {
            throw WriterControlFailure.invalidTransition
        }
        guard !bytes.isEmpty else { throw WriterControlFailure.inputEmpty }
        guard bytes.count <= ControllerWriterWireCodec.maxInputChunkBytes else {
            throw WriterControlFailure.inputChunkTooLarge
        }
        if kind == .paste, pasteRequiresConfirmation(bytes), !confirmed {
            throw WriterControlFailure.pasteConfirmationRequired
        }
        guard queue.count < Self.maxQueuedChunks else {
            throw WriterControlFailure.inputQueueFull
        }
        guard bytes.count <= Self.maxQueuedBytes - queuedBytes else {
            throw WriterControlFailure.inputQueueBytesExceeded
        }
        queue.append(PendingInput(commandID: commandID, bytes: bytes, kind: kind))
        queuedBytes += bytes.count
    }

    mutating func dequeue() -> PendingInput? {
        guard !queue.isEmpty else { return nil }
        let value = queue.removeFirst()
        queuedBytes -= value.bytes.count
        return value
    }

    mutating func setForeground(_ foreground: Bool) {
        isForeground = foreground
        if !foreground {
            lease = .lost
            clearPendingInput()
        }
    }

    private mutating func clearPendingInput() {
        queue.removeAll(keepingCapacity: true)
        queuedBytes = 0
    }
}
